string memoire
int32 val
